AK/BigIntBase.h

```diff
@@ -0,0 +1,595 @@
+/*
+ * Copyright (c) 2023, Dan Klishch <danilklishch@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/BuiltinWrappers.h>
+#include <AK/Span.h>
+#include <AK/StdLibExtras.h>
+#include <AK/Types.h>
+
+namespace AK {
+
+// Ideally, we want to store data in the native processor's words. However, for some algorithms,
+// particularly multiplication, we require double of the amount of the native word size.
+#if defined(__SIZEOF_INT128__) && defined(AK_ARCH_64_BIT)
+using NativeWord = u64;
+using DoubleWord = unsigned __int128;
+using SignedDoubleWord = __int128;
+#else
+using NativeWord = u32;
+using DoubleWord = u64;
+using SignedDoubleWord = i64;
+#endif
+
+template<bool sign>
+using ConditionallySignedDoubleWord = Conditional<sign, SignedDoubleWord, DoubleWord>;
+
+template<typename T>
+concept BuiltInUFixedInt = OneOf<T, bool, u8, u16, u32, u64, unsigned long, unsigned long long, DoubleWord>;
+
+template<typename T>
+constexpr inline size_t bit_width = sizeof(T) * 8;
+
+constexpr size_t word_size = bit_width<NativeWord>;
+constexpr NativeWord max_word = ~static_cast<NativeWord>(0);
+static_assert(word_size == 32 || word_size == 64);
+
+// Max big integer length is 256 MiB (2.1e9 bits) for 32-bit, 4 GiB (3.4e10 bits) for 64-bit.
+constexpr size_t max_big_int_length = 1 << (word_size == 32 ? 26 : 29);
+
+// ===== Static storage for big integers =====
+// FIXME: remove once Clang formats these properly.
+// clang-format off
+template<typename T, typename WordType = NativeWord>
+concept IntegerStorage = requires(T storage, size_t index)
+{
+    { storage.is_negative() } -> SameAs<bool>;
+    { storage.size() } -> SameAs<size_t>;
+    { storage[index] } -> ConvertibleTo<WordType&>;
+    { storage.data() } -> ConvertibleTo<WordType*>;
+};
+
+template<typename T>
+concept IntegerReadonlyStorage = IntegerStorage<T, NativeWord const>;
+// clang-format on
+
+template<size_t total_size>
+struct StackBuffer {
+    NativeWord data[total_size];
+    NativeWord* free_ptr = data;
+
+    constexpr NativeWord* allocate(size_t size)
+    {
+        NativeWord* result = free_ptr;
+        free_ptr += size;
+        VERIFY(free_ptr <= data + total_size);
+        return result;
+    }
+};
+
+template<>
+struct StackBuffer<0> {
+    NativeWord* allocate(size_t) { VERIFY_NOT_REACHED(); }
+};
```

Comment:
this is essentially a `std::monotonic_buffer_resource`. Are we sure that there's nowhere else in the project that's using a similar stack-based bump allocator?

Reply:
Oh, after countless refactors, `StackBuffer` became unused... I've just deleted it